Add an `undef` global Jinja function

lilatomic opened the pull request:
##### SUMMARY
Factors in the feedback on the previous attempt #75407 ; I couldn't reopen the PR. This PR originally made a filter, but we felt that wasn't really appropriate given the semantics (it doesn't actually filter anything). I think explicitly calling it Undefined also clarifies its semantics; the fail filter would just return `Undefined` so it didn't explode things, but people might have expected it to behave almost like the `fail` module.

Fixes #69255

Enables the creation of `Undefined` values. This is useful for specifying variables that must be overridden. These variables don't require an intermediate filtering and local variable For example, in a role's defaults file:
```yaml
global_mysql_root_user: "sqladmin"
global_mysql_root_password: "{{ undef(hint='required, generate with pwgen -s 20') }}"
```
instead of the suggested workaround
```yaml
global_mysql_root_user: "sqladmin"
# global_mysql_root_password: 
_global_mysql_root_password: "{{ global_mysql_root_password | mandatory('required, generate with pwgen -s 20') }}"
```
This makes it easier to search a codebase for where a variable is used, as it doesn't require also looking for all the local variants of that variable.

###### Alternatives

One suggestion was to extend the "mandatory" filter to treat `null` values as undefined. But there are valid use cases for requiring a user to specify a value but where `null` is a valid value.

Another suggestion was to use `validate_argument_spec`. I like this suggestion, argspec looks cool. It currently lacks documentation. As well, there are things like vars_files where the `fail` filter is much simpler than an argspec.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
ansible.template.Templar

##### ADDITIONAL INFORMATION
###### Basic Usage
If not overridden:
```bash
ansible localhost -m debug -a msg="{{ foo }}" -e foo="{{ undef('xfail') }}"
localhost | FAILED! => {
    "msg": "The task includes an option with an undefined variable. The error was: {{ undef('xfail') }}: xfail. {{ undef('xfail') }}: xfail"
}

ansible localhost -m debug -a msg="{{ foo }}" -e foo="{{ undef('xfail') }}" -e foo="hi"
localhost | SUCCESS => {
    "msg": "hi"
}
```
###### Other Usage:

```yaml
- name: demo
  hosts: localhost
  connection: local
  gather_facts: false

  tasks:
    - name: print mandatory
      debug:
        var: v0
      ignore_errors: true

    - name: "print fail"
      debug:
        var: v1

    - name: "print fail with default"
      debug:
        var: v2
      ignore_errors: true

    - name: "print fail in ternary"
      debug:
        var: v3
      loop: [True, False]
      ignore_errors: true

    - name: "print msg undef"
      debug:
        msg: "{{ v4 }}"
      ignore_errors: true
    - name: "inspect undef"
      debug:
        var: undefined
      ignore_errors: true

    - name: "undef is undefined"
      assert:
        that: undef() is undefined

  vars:
    v0: "{{ undef() | mandatory('xfail') }}"
    v1: "{{ undef(hint='xfail') }}"
    v2: "{{ undef('xfail') | default('xx') }}"
    v3: "{{ item | ternary(0, undef('xfail')) }}"
    v4: "{{ undef() }}"

```

yields:
```bash
ansible-playbook demo.yml -vvv

PLAYBOOK: demo.yml **************************************************************************************************************************
1 plays in demo.yml

PLAY [demo] *********************************************************************************************************************************
META: ran handlers

TASK [print mandatory] **********************************************************************************************************************
task path: /home/lilatomic/GIT/ansible/demo.yml:7
fatal: [localhost]: FAILED! => {
    "msg": "An unhandled exception occurred while templating '{{ undef() | mandatory('xfail') }}'. Error was a <class 'ansible.errors.AnsibleFilterError'>, original message: xfail"
}
...ignoring

TASK [print fail] ***************************************************************************************************************************
task path: /home/lilatomic/GIT/ansible/demo.yml:12
ok: [localhost] => {
    "v1": "VARIABLE IS NOT DEFINED!: {{ undef(hint='xfail') }}: xfail"
}

TASK [print fail with default] **************************************************************************************************************
task path: /home/lilatomic/GIT/ansible/demo.yml:16
ok: [localhost] => {
    "v2": "xx"
}

TASK [print fail in ternary] ****************************************************************************************************************
task path: /home/lilatomic/GIT/ansible/demo.yml:21
ok: [localhost] => (item=True) => {
    "ansible_loop_var": "item",
    "item": true,
    "v3": "0"
}
ok: [localhost] => (item=False) => {
    "ansible_loop_var": "item",
    "item": false,
    "v3": "VARIABLE IS NOT DEFINED!: {{ item | ternary(0, undef('xfail')) }}: xfail"
}

TASK [print msg undef] **********************************************************************************************************************
task path: /home/lilatomic/GIT/ansible/demo.yml:27
fatal: [localhost]: FAILED! => {
    "msg": "The task includes an option with an undefined variable. The error was: {{ undef() }}: Mandatory variable has not been overridden\n\nThe error appears to be in '/home/lilatomic/GIT/ansible/demo.yml': line 27, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n    - name: \"print msg undef\"\n      ^ here\n"
}
...ignoring

TASK [inspect undef] ************************************************************************************************************************
task path: /home/lilatomic/GIT/ansible/demo.yml:31
ok: [localhost] => {
    "undefined": "VARIABLE IS NOT DEFINED!: 'undefined' is undefined"
}

TASK [undef is undefined] *******************************************************************************************************************
task path: /home/lilatomic/GIT/ansible/demo.yml:36
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}
META: ran handlers
META: ran handlers

PLAY RECAP **********************************************************************************************************************************
localhost                  : ok=7    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=2   

```

lilatomic committed:
add tests for fail filter

also tests that fail does not block inspectability

lilatomic committed:
add fail filter

fallback message is a bit clunky,
since you can't invoke a filter without specifying an input.
That is, "{{ fail }}" doesn't work,
so you have to do "{{ None | fail }}"

lilatomic committed:
document 'fail' filter

lilatomic committed:
add changelog fragment

lilatomic committed:
fail filter uses default message on Undefined or emptystring

makes it slightly easier to use the default message:
```diff
- "{{ None | fail }}"
+ "{{ '' | fail }}"
```

and the user sees a slightly more relevant message
if the message itself is undefined:

```diff
- The error was: {{ failmsg | fail }}: 'failmsg' is undefined
+ The error was: {{ failmsg | fail }}: Mandatory variable has not been overridden
```

nitzmahone commented:
Sorry for the delay, haven't forgotten about this... We've kicked this around internally a bit more- curious your thoughts about doing this as a Jinja global function instead of a filter... A filter just doesn't feel "right" to most of us, since this isn't really a transformative operation- it's more a way of saying "I want this value to act undefined but with a user-supplied hint on access". So something like:

```
x: undefined_hint('x should be defined by frob xyz')
```

(configured here): https://github.com/ansible/ansible/blob/devel/lib/ansible/template/__init__.py#L692-L695

The naming is always also a challenge- `fail` is only sort-of-correct since it doesn't really fail, it's just undefined, so as you've noted, any usual way of recovering undefined works the same way... Open to ideas!

lilatomic committed:
rebuild as the builtin `Undefined`

lilatomic commented:
So, your example with "undefined_hint" got me thinking that `jinja2.runtime.Undefined` already _has_ a hint, and it's even used in this filter, and maybe what we're after is just the ability to create Undefined?

```yaml
x: "{{ Undefined(hint='x should be defined by frob xyz' }}"
```
And it makes perfect sense to me to add that as a Jinja global function. 
You can still not supply the hint if it's pretty obvious: `x: "{{ Undefined() }}"` => `

But it's maybe a bit easy to misuse: `x: "{{ Undefined }}"` => `"msg": "<bound method Templar._make_undefined of <ansible.template.Templar object at 0x7f6e2038a4c0>>"`


And if we feel _super clever_, we could even pass the class as the `Callable`. Upside: exposes the whole `Undefined` API; downside: is maybe _too clever_.
```python
        self.environment.globals['Undefined'] = AnsibleUndefined
```

At first I was reluctant to add a Jinja global function. I'm having trouble articulating why. Maybe it feels a bit like cheating when the designers of a programming language use secret tools that the users can't hook into. Not that Ansible _needs_ another category of plugins, but if we're finding a use for this type of global, what about others? We have the convenience of the `now` global, what about a `uuid4` global (or one scoped by collection)? But reframing it as exposing `Undefined`, a "platform" feature, seems to counter that for me. 

I'm pushing the change, let me know your thoughts. I'm not sure where the doc goes now. I've left it where it is, since next to `mandatory` is where I would look for how to do this and it's filter-like enough. I'll update the sample at the top.

lilatomic commented:
Also I called it "Undefined" since that's the base class of the hierarchy. But I have this sinking feeling that this will confuse _something_ out there which is using reflection.

nitzmahone commented:
Yeah, I had a bunch of similar thoughts while kicking this one around. I think I'm happy with everything in your current impl except the name of the function, as I share the same concerns around collisions or potential future effects in Jinja itself. I'm not sure how much of the underlying undefined capability we'd want to expose through Ansible ("YAGNI" and all ;) ), especially given the wide range of Jinja versions typically in use. We're looking at a pretty big bump there in the next release or so, but still probably keeping it to just the things we know are needed is good.

lilatomic commented:
I looked through the main Ansible codebases using the [codesearch](https://codesearch.community.eng.ansible.com/). 
- "Undefined" _seems_ safe, but is the same as jinja's `Undefined`.
- "undefined" is used in jinja tests and separately in ansible `:/lib/ansible/playbook/conditional.py`. "undefined" is _also_ used as a valid parameter in some modules(eg [here](https://docs.ansible.com/ansible/latest/collections/ovirt/ovirt/ovirt_cluster_module.html#parameter-cpu_arch) ). So calling it this could break things
-  "Undef" seems available
- "undef" is used as a valid parameter in some modules (eg [here](https://docs.ansible.com/ansible/latest/collections/netapp/ontap/na_ontap_interface_module.html#parameter-role) ). So calling it this could break things

I think Undef is fine

lilatomic commented:
oh also, I'm not too sure what the changelog fragment should look like. I think this is a minor change since it's not a filter?

samccann commented on docs/docsite/rst/user_guide/playbooks_filters.rst:
```diff
@@ -68,6 +68,11 @@ If you configure Ansible to ignore undefined variables, you may want to define s
 
 The variable value will be used as is, but the template evaluation will raise an error if it is undefined.
 
+A convenient way of requiring a variable to be overridden is to give it an undefined value using the ``Undef`` keyword. This can be useful in a role's defaults::
```
Please use a code-block here instead of :: .  The translation of Ansible docs can ignore content within a code-block but has problems with ::.  See https://docs.ansible.com/ansible/latest/dev_guide/style_guide/index.html#syntax-highlighting-pygments for an example of a code-block.

lilatomic replied:
I had copypasted from another section in that file. Is it a problem that many of the other segments don't use a code-block?

samccann replied:
It makes things harder for translation. If you feel like changing the rest, sure, but don't feel you have to.

s-hertel commented on docs/docsite/rst/user_guide/playbooks_filters.rst:
```diff
@@ -68,6 +68,11 @@ If you configure Ansible to ignore undefined variables, you may want to define s
+A convenient way of requiring a variable to be overridden is to give it an undefined value using the ``Undef`` keyword. This can be useful in a role's defaults::
+
+    galaxy_url: "https://galaxy.ansible.com"
+    galaxy_api_key: {{ Undef(hint="You must specify your Galaxy API key") }}
```
The docs/tests use `hint`, but the code uses `msg`.

lilatomic committed:
harmonise `hint` parameter for make_undefined with jinja

lilatomic committed:
use code block for documentation item

[ref](https://github.com/ansible/ansible/pull/75435#discussion_r707661035)

nitzmahone reviewed (changes requested):
OK, one last change to request and I think we're good to merge this- we've been waffling on whether to "leak" the Python convention of naming a factory callable with a capital letter into Jinja, and we landed (just barely) on "don't"... So if you can make it `undef` instead of `Undef` (or if you're OK with one of us making the change for you), we can get this merged and included in 2.12. Thanks for sticking with us on this one!

lilatomic committed:
rename to `undef` to expose less Python into the Jinja

[ref](https://github.com/ansible/ansible/pull/75435#pullrequestreview-757799031)

lilatomic committed:
explicitly instantiate undefined value now that it's possible

see I knew we would break something with reflection

lilatomic commented:
So, it looks like renaming it `undef` broke some things. In the tests, it's only been people trying to instantiate undefined variables, so it's a simple enough fix: `undef` -> `undef()`. So I think it's a good indication that people developing the Ansible codebase were expecting `undef` to give them undefined values. It does mean that this might break other people's code which expects `undef` to be undefined, rather than `"<bound method Templar._make_undefined of <ansible.template.Templar object at 0x7ff17a2fa490>>"`

nitzmahone committed:
preserve test coverage of undefined variable

lilatomic commented:
oh, I made the test test for the value of Undefined and not an undefined value. Good catch!

nitzmahone commented:
> oh, I made the test test for the value of Undefined and not an undefined value. Good catch!

Yeah, the new behavior is a good thing to add to the original test as well, but definitely also want to ensure that an arbitrary undefined value is also `AnsibleUndefined` as the test did (eg, to ensure that our chained undefined stuff is working instead of blowing up as it would with Jinja-native `Undefined`).

For the original question: there was nothing special about `undef`- we could technically have a collision like that with existing content anytime a new global name gets reserved, so I'm not too worried about it. Thanks for all the shuffle on this one- I'm hitting merge!
